<a href="https://colab.research.google.com/github/fotojuanmoreno/DescubriendoML/blob/master/3.-regresion_lineal_1/RegresionLineal1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Utilizaremos la base de boston que es un set de datos sobre los precios de las viviendas en boston en relación a ditintos factores como las habitaciones que tienen o el indice de criminalidad de la zona.

Cargamos los modulos:

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge

Cargamos nuestra base de datos y le la asignamos a nuestra variable boston:


In [ ]:
boston = load_boston()

Utilizamos la propiedad keys() para ver las distintas claves de nuestra db:


In [ ]:
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

Veamos un poco la informacion que contiene boston segun la información que nos ha dado boston.keys():

In [ ]:
boston.data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [ ]:
boston.target

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
       18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
       15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
       13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
       21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
       35.4, 24.7, 31.6, 23.3, 19.6, 18.7, 16. , 22.2, 25. , 33. , 23.5,
       19.4, 22. , 17.4, 20.9, 24.2, 21.7, 22.8, 23.4, 24.1, 21.4, 20. ,
       20.8, 21.2, 20.3, 28. , 23.9, 24.8, 22.9, 23.9, 26.6, 22.5, 22.2,
       23.6, 28.7, 22.6, 22. , 22.9, 25. , 20.6, 28.4, 21.4, 38.7, 43.8,
       33.2, 27.5, 26.5, 18.6, 19.3, 20.1, 19.5, 19.5, 20.4, 19.8, 19.4,
       21.7, 22.8, 18.8, 18.7, 18.5, 18.3, 21.2, 19.2, 20.4, 19.3, 22. ,
       20.3, 20.5, 17.3, 18.8, 21.4, 15.7, 16.2, 18. , 14.3, 19.2, 19.6,
       23. , 18.4, 15.6, 18.1, 17.4, 17.1, 13.3, 17.8, 14. , 14.4, 13.4,
       15.6, 11.8, 13.8, 15.6, 14.6, 17.8, 15.4, 21

In [ ]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [ ]:
boston.DESCR

".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000

In [ ]:
boston.filename

'/usr/local/lib/python3.6/dist-packages/sklearn/datasets/data/boston_house_prices.csv'

boston.DESCR nos ha dado una información relevante. Nos indica que el numero de instancias es de 506, y el de atributos es 13. Estos dos elementos podemos comprobarlos con:

In [ ]:
boston.data.shape

(506, 13)

Visto que es cierto, veamos si <i>target</i> contiene las 506 respuestas:

In [ ]:
boston.target.shape

(506,)

Comprobado que todo está correcto, preparemos nuestros datos repartiendolos entre los que utilizaremos para el entrenamiento y los que servirán para probarlo.
Para esto uliizaremos el metodo train_test_split que hemos traido de sklearn.model_selection que repartirá nuestros datos:

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(boston.data, boston.target)

podemos comprobar que esto ha salido bien viendo los datos que comtiene X_train, por ejemplo:

In [ ]:
X_train.shape

(379, 13)

Bien, ahora utilizaremos el KNeighborsRegressor para indicarle a nuestro código que tenga en cuenta los valores de los 3 elementos más cernanos a la hora de situar o inspeccionar un elemento:

In [ ]:
knn = KNeighborsRegressor(n_neighbors = 3)

Y ya podemos comenzar el entrenamiento:

In [ ]:
knn.fit(X_train, Y_train)


KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
                    metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                    weights='uniform')

Veamos lo que ha aprendido nuestro algoritmo. Utilizando los valores que hemor preparado para las pruebas (X_test, Y_test) el programa intentará averiguar cual deberia de ser "su lugar". Con score se nos devolvera el porcentaje de aciertos que haya habido y así sabremos como está funcionando el programa:

In [ ]:
knn.score(X_test, Y_test)

0.43765229007656226

Es facil adivinar que el resultado es muy malo. Podemos probar a realizar el mismo ejercicio pero aumentando el número de vecinos que tendrá en cuenta. Yo he estado probando un poco y, tristemente este es el mejor resultado que he obtedido. Podriais probar a realizar un bucle for con un range para ver los distintos valores que os puede dar como resultado (si no son excesibos los datos de entrada, claro)

In [ ]:
knn = KNeighborsRegressor(n_neighbors = 10)
knn.fit(X_train, Y_train)

KNeighborsRegressor(algorithm='auto', leaf_size=30, metric='minkowski',
                    metric_params=None, n_jobs=None, n_neighbors=10, p=2,
                    weights='uniform')

In [ ]:
knn.score(X_test, Y_test)

0.44573662941537623

Ahora vamos a borran nuestra variable knn y realizaremos el mismo proceso mediante regresión lineal:

In [ ]:
del knn

In [ ]:
rl = LinearRegression()

In [ ]:
rl.fit(X_train, Y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

y al igual que antes, comprobemos que tal ha ido el entrenamiento:

In [ ]:
rl.score(X_test, Y_test)

0.7554423626546778

Parece que mucho mejor.

Ahora probemos con el algoritmo Ridge. Como antes, primero borraremos este modelo (es solo para evitar saturar la máquina con demasiados modelos):

In [ ]:
del rl

In [ ]:
ridge = Ridge()

In [ ]:
ridge.fit(X_train, Y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [ ]:
ridge.score(X_test, Y_test)

0.7498866405679339

Este ejercicio no ha explicado mucho pero nos enseña un poco como distintas formas de hacer lo mismo dan distintos resultados. Y esto no tiene que significar necesariamente que unos algoritmos funcionen mejor que otros, simplemente admiten distintas configuraciones (que no hemos usado en general) y dependiendo de nuestros datos un método se adaptará mejor que otro.
